File: 算法笔记/NQSDemo/src/main/java/jeason/a线性表/a动态数组/ArrayList.java
package jeason.a线性表.a动态数组;

@SuppressWarnings("unchecked")
public class ArrayList<E> {
	//元素的数量
	private int size;
	//所有的元素
	private E[] elements;
	
	//默认数组分配内存空间
	private static final int DEFAULT_CAPACITY = 10;
	private static final int ELEMENT_NOT_FOUND = -1;
	

	public ArrayList(int capaticy) {
		capaticy = (capaticy < DEFAULT_CAPACITY) ? DEFAULT_CAPACITY : capaticy;
		elements = (E[]) new Object[capaticy];
	}
	
	public ArrayList() {
		this(DEFAULT_CAPACITY);
	}
	
	/** 清除所有元素 */
	public void clear() {
		for (int i = 0; i < size; i++) {
			elements[i] = null;
		}
		size = 0;
	}

	/** 元素的数量 */
	public int size() {
		return size;
	}

	/** 是否为空 */
	public boolean isEmpty() {
		return size == 0;
	}

	/** 是否包含某个元素 */
	public boolean contains(E element) {
		return indexOf(element) != ELEMENT_NOT_FOUND;
	}

	/** 添加元素到尾部 */
	public void add(E element) {
		add(size, element);
	}

	/** 获取index位置的元素 */
	public E get(int index) {
		rangeCheck(index);
		return elements[index];
	}

	/** 设置index位置的元素 */
	public E set(int index, E element) {
		rangeCheck(index);
		
		E old = elements[index];
		elements[index] = element;
		return old;
	}

	/** 在index位置插入一个元素 */
	public void add(int index, E element) {
		rangeCheckForAdd(index);
		
		ensureCapacity(size + 1);
		
		for (int i = size; i > index; i--) {
			elements[i] = elements[i - 1];
		}
		elements[index] = element;
		size++;
	}

	/** 删除index位置的元素 */
	public E remove(int index) {
		rangeCheck(index);
		
		E old = elements[index];
		for (int i = index + 1; i < size; i++) {
			elements[i - 1] = elements[i];
		}
		//最后一个元素置null，不然指针没释放，指针指向的内存地址也没释放
		elements[--size] = null;
		return old;
	}

	/** 查看元素的索引 */
	public int indexOf(E element) {
		if (element == null) {  // 1
			for (int i = 0; i < size; i++) {
				if (elements[i] == null) return i; 
			}
		} else {
			for (int i = 0; i < size; i++) {
				//java里面equals是比较内存地址是否相同
				if (element.equals(elements[i])) return i; // n
			}
		}
		return ELEMENT_NOT_FOUND;
	}

	
	/** 保证要有capacity的容量 */
	private void ensureCapacity(int capacity) {
		int oldCapacity = elements.length;
		if (oldCapacity >= capacity) return;
		
		// 新容量为旧容量的1.5倍
		int newCapacity = oldCapacity + (oldCapacity >> 1);
		E[] newElements = (E[]) new Object[newCapacity];
		for (int i = 0; i < size; i++) {
			newElements[i] = elements[i];
		}
		elements = newElements;
		
		System.out.println(oldCapacity + "扩容为" + newCapacity);
	}
	//打印抛出异常
	private void outOfBounds(int index) {
		throw new IndexOutOfBoundsException("Index:" + index + ", Size:" + size);
	}
	
	//监测index是否越界
	private void rangeCheck(int index) {
		if (index < 0 || index >= size) {
			outOfBounds(index);
		}
	}
	//监测index是否越界
	private void rangeCheckForAdd(int index) {
		if (index < 0 || index > size) {
			outOfBounds(index);
		}
	}
	
	@Override
	public String toString() {
		// size=3, [99, 88, 77]
		StringBuilder string = new StringBuilder();
		string.append("size=").append(size).append(", [");
		for (int i = 0; i < size; i++) {
			if (i != 0) {
				string.append(", ");
			}
			
			string.append(elements[i]);
			
		}
		string.append("]");
		return string.toString();
	}
}

File: PythonAlgorithm/Study/a线性表/a动态数组/ArrayList.py
from typing import Generic, TypeVar, List
E = TypeVar('E')


class Person(object):
    def __init__(self, name, age) -> None:
        self.name = name
        self.age = age
        
    def toString(self) -> str:
        print(f"Person:[name={self.name} age={self.age}]")
    
    def __eq__(self, __value: object) -> bool:
        if object and (type(__value) == type(self)):
            return (self.name == __value.name) and (self.age == __value.age)



# 定义ArrayListClass为泛型，根据实例化时传入的类型决定
# int_list = ArrayListClass[int]()
# int_list.append(1)
#
# string_list = ArrayListClass[str]()        
# string_list.append("Hello")
class ArrayListClass(Generic[E]):

    
    def __init__(self, *, capacity: int) -> None:
        
        DEFAULT_CAPACITY: int = 10 # 默认数组分配内存空间
        
        # 容量不能小于默认容量(10)
        self.capacity = capacity if (capacity >= DEFAULT_CAPACITY) else DEFAULT_CAPACITY
        # 动态数组
        self.__elements: List[E] = [None] * capacity
        # 元素数量
        self.__eleCount: int = 0

    def clear(self) -> None:
        # 清除所有元素
        for i in range(self.__eleCount):
            self.__elements[i] = None
        self.__eleCount = 0
        self.__elements = None
        self.capacity = 0
    
    def caculateEleCount(self) -> int:
        # 元素的数量
        return self.__eleCount

    def is_empty(self) -> bool:
        # 是否为空
        return self.__eleCount == 0

    def contains(self, element: E) -> bool:
        # 是否包含某个元素
        return self.find(element) != -1

    def addToTail(self, element: E) -> None:
        # 添加元素到尾部
        self.insertObjectAtIndex(self.__eleCount, element)

    def getObjectAtIndex(self, index: int) -> E:
        # 获取index位置的元素
        self.__range_check(index)
        return self.__elements[index]

    def setObjectAtIndex(self, index: int, element: E) -> None:
        # 设置index位置的元素
        self.__range_check(index)
        self.__elements[index] = element
        

    def insertObjectAtIndex(self, index: int, element: E) -> None:
        # 在index位置插入一个元素
        self.__range_check_for_add(index)
        self.__ensure_capacity(self.__eleCount + 1)
        
        for i in range(self.__eleCount, index, -1):
            self.__elements[i] = self.__elements[i - 1]
        
        self.__elements[index] = element
        self.__eleCount += 1

    def removeObjectAtIndex(self, index: int) -> E:
        # 删除index位置的元素
        self.__range_check(index)
        old = self.__elements[index]
        
        for i in range(index + 1, self.__eleCount):
            self.__elements[i - 1] = self.__elements[i]
        
        # 将删除前的最后一个元素置空
        self.__elements[self.__eleCount - 1] = None
        
        self.__eleCount -= 1
        
        # 动态缩容
        self.__trim()
        return old

    def indexOfObject(self, element: E) -> int:
        # 查看元素的索引
        if element is None:
            for i in range(self.__eleCount):
                if self.__elements[i] is None:
                    return i
        else:
            for i in range(self.__eleCount):
                if element == self.__elements[i]:
                    return i
        return -1

    def __ensure_capacity(self, tempCapacity: int) -> None:
        # 保证要有capacity的容量
        if self.capacity > tempCapacity:
            return
        
        old_capacity = self.capacity
        new_capacity = 0
        # 新容量为旧容量的1.5倍
        if old_capacity == 0:
            new_capacity = 10
        else:
            new_capacity = int(self.capacity * 1.5)
        
        new_elements = [None] * new_capacity
        for i in range(self.__eleCount):
            new_elements[i] = self.__elements[i]
        
        self.__elements = new_elements
        self.capacity = new_capacity
        print(f"数组容量从:{old_capacity} 扩容到: {new_capacity}")

    def __trim(self):
        # 30
        oldCapacity: int  = len(self.__elements)
        # 15
        newCapacity: int  = int(oldCapacity / 2)
        if ((self.__eleCount > newCapacity) or (oldCapacity < 10)):
            return

        # 当前可以进行数组缩容
        new_elements = [None] * newCapacity
        for i in range(self.__eleCount):
            new_elements[i] = self.__elements[i]
        self.__elements = new_elements
        print(f"数组容量从:{oldCapacity} 缩容到: {newCapacity}")

    def __out_of_bounds(self, index: int) -> None:
        # 打印抛出异常
        raise IndexError(f"下标越界：Index: {index}, eleCount: {self.__eleCount}")

    def __range_check(self, index: int) -> None:
        # 监测index是否越界
        if index < 0 or index >= self.__eleCount:
            self.__out_of_bounds(index)

    def __range_check_for_add(self, index: int) -> None:
        # 监测index是否越界 
        if index < 0 or index > self.__eleCount:
            self.__out_of_bounds(index)

    def __str__(self) -> str:
        string = ["动态数组汇总：\n\t元素数量:" + str(self.__eleCount) + ", \n\t"]
        for i in range(self.__eleCount):
            if i != 0:
                string.append("\n\t")
            string.append(f"index:{i}："+str(self.__elements[i]))
        return "".join(string)
    


